<a href="https://colab.research.google.com/github/tutsilianna/Automatic_Text_Processing_and_Image_Processing/blob/main/Introduction%20to%20Automatic%20Text%20Processing/Task_1_%7C_Introduction_to_Automatic_Text_Processing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Text preprocessing

**Data URL**

In [1]:
DATA_URL = "https://www.gutenberg.org/files/913/913-0.txt"

In [2]:
! pip install -q nltk==3.2.5

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 8.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done


Loading *NLTK*'s 'wordnet'

In [3]:
import nltk
from nltk.corpus import wordnet
nltk.download("wordnet")

lemmatizer = nltk.stem.wordnet.WordNetLemmatizer()
lemmatizer

[nltk_data] Downloading package wordnet to /root/nltk_data...


<WordNetLemmatizer>

Downloading the text for the task via `urllib`

In [4]:
import urllib.request

opener = urllib.request.URLopener({})
resource = opener.open(DATA_URL)
charset = resource.headers.get_content_charset()
print("Charset", charset)
raw_text = resource.read()

if charset:
  raw_text = raw_text.decode(resource.headers.get_content_charset())
else:
  raw_text = raw_text.decode("utf-8")

raw_text[:100]

Charset None


'\ufeffThe Project Gutenberg EBook of A Hero of Our Time, by M. Y. Lermontov\r\n\r\nThis eBook is for the use '

Removing the book ending (Gutenberg legal information)

In [5]:
import re

clean_pattern = re.compile("End of the Project Gutenberg EBook.*")
cleaner_text =  re.sub(clean_pattern, "", raw_text.replace("\n", " ").replace("\r", " "))
cleaner_text[-100:]

'ases, equivalent to: “Men are fools, fortune is  blind, and life is not worth a straw.”]            '

Splitting the text into tokens with a little help from [NLTK](https://nltk.org/).

In [6]:
from nltk.tokenize import word_tokenize
import nltk
nltk.download('punkt')

! unzip /root/nltk_data/corpora/wordnet.zip -d /root/nltk_data/corpora/;

tokens =  word_tokenize(cleaner_text)
"A total of %d 'tokens'" % len(tokens)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
Archive:  /root/nltk_data/corpora/wordnet.zip
   creating: /root/nltk_data/corpora/wordnet/
  inflating: /root/nltk_data/corpora/wordnet/lexnames  
  inflating: /root/nltk_data/corpora/wordnet/data.verb  
  inflating: /root/nltk_data/corpora/wordnet/index.adv  
  inflating: /root/nltk_data/corpora/wordnet/adv.exc  
  inflating: /root/nltk_data/corpora/wordnet/index.verb  
  inflating: /root/nltk_data/corpora/wordnet/cntlist.rev  
  inflating: /root/nltk_data/corpora/wordnet/data.adj  
  inflating: /root/nltk_data/corpora/wordnet/index.adj  
  inflating: /root/nltk_data/corpora/wordnet/LICENSE  
  inflating: /root/nltk_data/corpora/wordnet/citation.bib  
  inflating: /root/nltk_data/corpora/wordnet/noun.exc  
  inflating: /root/nltk_data/corpora/wordnet/verb.exc  
  inflating: /root/nltk_data/corpora/wordnet/README  
  inflating: /root/nltk_data/corpora/wordnet/index.sense  
  infla

"A total of 72549 'tokens'"

Now we are about to **lemmatize the tokens**. Please note that for better results we should have first PoS-tagged the text (e.g. with NLTK as well, [please refer to the book and the docs](https://www.nltk.org/book/ch05.html)). `WordNetLemmatizer` would work best with PoS tags provided. However, to make things short and simple, we won't do it as of now.



---


## Task \#1
using Python's standard library's `str.isalpha` modify the code below to remove all non-letter tokens from sentences.

---



In [7]:
from tqdm.notebook import tqdm

lemmas = [lemmatizer.lemmatize(lemma) for lemma in tqdm(tokens) if lemmatizer.lemmatize(lemma).isalpha()]
lemmas[-10:]

  0%|          | 0/72549 [00:00<?, ?it/s]

['fortune', 'is', 'blind', 'and', 'life', 'is', 'not', 'worth', 'a', 'straw']

Let us check if everything is correct. Total corpus size should be equal to 59087

In [8]:
len(lemmas)

59087

To continue working on this assignment make sure those numbers match!


---

## Task #2

Using `lemmas`, english NLTK stopwords (`nltk.corpus.stopwords`) and `nltk.FreqDist`, compute **the fraction of the stopwords** in the top-100 most frequent words in the text.

E.g. if there were just **3 stopwords in 100** most frequent words in the text, the answer would be **0.03**

**Googling how to work with NLTK is encouraged.**

---



In [10]:
import nltk
from nltk import FreqDist
from nltk.corpus import stopwords
nltk.download("stopwords")
STOPWORDS = set(stopwords.words("english"))
stopwords.words("english")[:5]

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


['i', 'me', 'my', 'myself', 'we']

In [36]:
most_common100 = FreqDist(lemmas).most_common(100)
most_common100[:5]

[('the', 2937), ('I', 1802), ('a', 1729), ('to', 1678), ('and', 1528)]

In [40]:
print(f'The fraction of the stopwords in the top-100 most frequent words in the text {sum([1 for word, _ in most_common100 if word in stopwords.words("english")])/100}')

The fraction of the stopwords in the top-100 most frequent words in the text 0.66


Check yourself: stopwords rate in 100 most frequent words in the text should be 0.66



---


## Task #3
Compute how many tokens occur in the text **strictly greater than** 50 times.

---



In [51]:
print(f'How many tokens occur in the text strictly greater than 50 times? {sum([1 for _, count in FreqDist(lemmas).items() if count > 50])}')

How many tokens occur in the text strictly greater than 50 times? 149


Chek yourself: 149



---


## Individual task
Click [the link](https://www.gutenberg.org/files/2142/2142-h/2142-h.htm) to see an example of the Russian classical literature. Your task is to analyze the text.

* Enter the number of letter tokens
* What is the share of stop-words of 150 most frequent tokens in the text?
* What is the amount of tokens encountered in the text strictly greater than 20 times?
---

In [52]:
DATA_URL = 'https://www.gutenberg.org/files/2142/2142-0.txt'

Downloading the text for the task via `urllib`

In [65]:
opener = urllib.request.URLopener({})
resource = opener.open(DATA_URL)
charset = resource.headers.get_content_charset()
print("Charset", charset)
raw_text = resource.read()

if charset:
  raw_text = raw_text.decode(resource.headers.get_content_charset())
else:
  raw_text = raw_text.decode("utf-8")

raw_text[:100]

Charset None


'\ufeffThe Project Gutenberg EBook of Childhood, by Leo Tolstoy\r\n\r\nThis eBook is for the use of anyone any'

Removing the book ending (Gutenberg legal information)

In [74]:
import re

clean_pattern = re.compile("End of the Project Gutenberg EBook.*")
cleaner_text =  re.sub(clean_pattern, "", raw_text.replace("\n", " ").replace("\r", " "))
cleaner_text[-100:]

'ence unite me to those two beings solely  in order to make me regret them my life long?”            '

Splitting the text into tokens with a little help from [NLTK](https://nltk.org/).

In [75]:
tokens =  word_tokenize(cleaner_text)
"A total of %d 'tokens'" % len(tokens)

"A total of 45787 'tokens'"

using Python's standard library's `str.isalpha` modify the code below to remove all non-letter tokens from sentences

In [76]:
lemmas = [lemmatizer.lemmatize(lemma) for lemma in tqdm(tokens) if lemmatizer.lemmatize(lemma).isalpha()]
lemmas[-10:]

  0%|          | 0/45787 [00:00<?, ?it/s]

['in', 'order', 'to', 'make', 'me', 'regret', 'them', 'my', 'life', 'long']

In [73]:
len(lemmas)

38587

In [64]:
most_common150 = FreqDist(lemmas).most_common(150)
most_common150[:5]

[('the', 1796), ('and', 1455), ('to', 1250), ('a', 1171), ('I', 955)]

In [63]:
print(f'The fraction of the stopwords in the top-150 most frequent words in the text {sum([1 for word, _ in most_common150 if word in stopwords.words("english")])/150}')

The fraction of the stopwords in the top-150 most frequent words in the text 0.54


In [62]:
print(f'How many tokens occur in the text strictly greater than 20 times? {sum([1 for _, count in FreqDist(lemmas).items() if count > 20])}')

How many tokens occur in the text strictly greater than 20 times? 248
